ArjunaJTA/cdi/tests/classes/com/hp/mwtests/ts/jta/cdi/transactional/TransactionalImplTest.java

```diff
@@ -139,6 +141,80 @@ public void testDefaultThrowException() throws Throwable {
         AssertionParticipant.assertCommitted();
     }
 
+    @Test
+    public void testDefaultThrowError() throws Throwable {
+        // java.lang.Error is un-checked exception and rollback is expected
+
+        Utills.assertTransactionActive(false);
+
+        try {
+            testTransactionalBean.invokeWithDefault(Error.class);
+            Assert.fail("Expected Throwable to be thrown and rollback the transaction, but it was not");
+        } catch (Throwable t) {
+            log.debug("Expected behaviour: Error was thrown from the method", t);
+        }
+
+        Utills.assertTransactionActive(false);
+        AssertionParticipant.assertRolledBack();
+    }
+
+    @Test
+    public void testThrowErrorWithDontRollbackOn() throws Throwable {
+
+        Utills.assertTransactionActive(false);
+
+        try {
+            testTransactionalBean.invokeWithRequiresAndDontRollbackOnError(Error.class);
+            Assert.fail("Expected Throwable to be thrown but not rolling back the transaction, but it was not thrown");
+        } catch (Throwable t) {
+            log.debug("Expected behaviour: Error was thrown from the method", t);
+        }
+
+        Utills.assertTransactionActive(false);
+        AssertionParticipant.assertCommitted();
+    }
+
+    @Test
+    public void testDefaultThrowErrorWhenAcquiringTxn() throws Exception {
+        // java.lang.Error is un-checked exception and rollback is expected
+
+        TransactionManager tm = Utills.getTransactionManager();
+        Utills.assertTransactionActive(false);
+        tm.begin();
+
+        try {
+            Utills.assertTransactionActive(true);
+            testTransactionalBean.invokeWithDefault(Error.class);
+            Assert.fail("Expected Throwable to be thrown and rollback the transaction, but it was not");
+        } catch (Throwable t) {
+            log.debug("Expected behaviour: Error was thrown from the method", t);
+        }
+        tm.rollback();
```

Comment:
The transaction should have already rolled back so this call means that we are testing that it did.

Reply:
That's not right. The interceptor does not rollback the transaction, it marks the transaction with `setRollbackOnly()`. It's the same as for EJB.
Calling the `rollback` method is the responsibility of the actor who started the transaction. When the transaction is started by interceptor then it's interceptor that is responsible to `rollback`. When the test starts the transaction then the test has to `rollback` it.

Reply:
Ah yes I forgot, excellent point.